sensor/common/clusterentities/store_endpoints.go

```diff
@@ -0,0 +1,328 @@
+	// Maps holding historical data must not be wiped on reset! Instead, all entities must be marked as historical.
+	// Must be called before the respective source maps are wiped!
+	// Performance optimization: no need to handle history if history is disabled
+	if !e.historyEnabled() {
+		e.initMapsNoLock()
+		return
+	}
+	e.moveAllToHistory()
+
+	e.endpointMap = make(map[net.NumericEndpoint]map[string]set.Set[EndpointTargetInfo])
+	e.reverseEndpointMap = make(map[string]set.Set[net.NumericEndpoint])
+	e.updateMetricsNoLock()
+}
+
+func (e *endpointsStore) historyEnabled() bool {
+	return e.memorySize > 0
+}
+
+func (e *endpointsStore) updateMetricsNoLock() {
+	metrics.UpdateNumberOfEndpoints(len(e.endpointMap), len(e.historicalEndpoints))
+}
+
+// RecordTick records a tick and returns all public IP addresses for which the count should be decremented
+func (e *endpointsStore) RecordTick() bool {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	removed := false
+	for endpoint, m1 := range e.historicalEndpoints {
+		for deploymentID, m2 := range m1 {
+			for _, status := range m2 {
+				status.recordTick()
+			}
+			e.reverseHistoricalEndpoints[deploymentID][endpoint].recordTick()
+			// Remove all historical entries that expired in this tick.
+			removed = e.removeFromHistoryIfExpired(deploymentID, endpoint) || removed
+		}
+	}
+	return removed
+}
+
+func (e *endpointsStore) Apply(updates map[string]*EntityData, incremental bool) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	e.applyNoLock(updates, incremental)
+}
+
+func (e *endpointsStore) applyNoLock(updates map[string]*EntityData, incremental bool) {
+	defer e.updateMetricsNoLock()
+	if !incremental {
+		for deploymentID := range updates {
+			e.purgeNoLock(deploymentID)
+		}
+	}
+	for deploymentID, data := range updates {
+		if data.isDeleteOnly() {
+			// A call to Apply() with empty payload of the updates map (no values) is meant to be a delete operation.
+			continue
+		}
+		e.applySingleNoLock(deploymentID, *data)
+	}
+}
+
+func (e *endpointsStore) purgeNoLock(deploymentID string) set.Set[net.NumericEndpoint] {
+	dec := set.NewSet[net.NumericEndpoint]()
+	// We will be manipulating reverseEndpointMap when calling deleteFromCurrent or moveToHistory,
+	// so let's make a temporary copy
+	endpointsSet := e.reverseEndpointMap[deploymentID]
+	for ep := range endpointsSet {
+		if e.historyEnabled() {
+			// When history expires, the counter will be decremented
+			e.moveToHistory(deploymentID, ep)
+		} else {
+			e.deleteFromCurrent(deploymentID, ep)
+			if ipAddr := ep.IPAndPort.Address; ipAddr.IsPublic() {
+				dec.Add(ep)
+			}
+		}
+	}
+	return dec
+}
+
+func (e *endpointsStore) applySingleNoLock(deploymentID string, data EntityData) {
+	dSet, deploymentFound := e.reverseEndpointMap[deploymentID]
+	if !deploymentFound || dSet == nil {
+		dSet = set.NewSet[net.NumericEndpoint]()
+	}
+
+	for ep, targetInfos := range data.endpoints {
+		dSet.Add(ep)
+		e.reverseEndpointMap[deploymentID] = dSet
```

Comment:
Yes, but only if `deploymentFound==true`, otherwise it cannot. I prefer to keep it here in this particular case (I removed it in the other place where this comment applied).